### ISSUE TITLE
The future of the project

### ISSUE THREAD

**Issue by tamasfe** [tamasfe/taplo#715]
**tl;dr**: I can't see myself returning here, and I'm asking you to decide the fate of this project.

When I started working on taplo, I wanted it to be a tool that puts TOML on par with YAML and JSON in terms of dev experience and perhaps even surpass them, in some aspects I think I achieved that and I'm happy that it ended up being useful for many. However the project was pretty much a learning experience for me, it was the first parser I have written, the first LSP, etc. and it really shows.

I still have a lot of plans, for example:

- rework the parser so that it is not a mess (e.g. I really liked the approach I ended up with in [rhai LSP](https://github.com/rhaiscript/lsp/blob/master/crates/rhai-rowan/src/parser/parsers/mod.rs), I think `rust-analyzer` has a similar architecture)
- adjust the code that builds on top of it so that it won't rely on an [accidentally introduced behavior that actually makes no sense](https://github.com/tamasfe/taplo/blob/a8bc571ee28775e7d5ad84c2ea87cf5b61ab42f5/crates/taplo/src/parser/mod.rs#L185)
- rethink/rework whatever the hell is happening after syntax parsing, to date I still have no idea how to efficiently and cleanly turn TOML into a DOM with its arbitrary rules and parent nodes appearing out of order
- rebase the LSP onto a maintained library, like [tower-lsp](https://github.com/ebkalderon/tower-lsp)
- rework the formatter so that it handles line breaks and nesting properly, something like [prettyplease](https://github.com/dtolnay/prettyplease) or [rhai-fmt](https://github.com/rhaiscript/lsp/tree/master/crates/rhai-fmt) (that was also based on prettyplease).
- fix filepath/URL handling, I think there is still confusion around this (unless the `windows<->everything else` conflicts were fixed while I was away)
- schemas and in general position querying (which TOML node maps to which property in schemas) is quite buggy
- ... probably many more (looking at the open issues)

I have practically disappeared from this project for the past 1-2 years, and now it is safe to say that whatever my plans are, they are very unlikely to be fulfilled. Nowadays I have a lot less free time, and even in that free time I code less as a hobby. On top of that being a maintainer really feels like a 2nd or 3rd job and I have a lot of respect for people who can pull it off, I definitely cannot.

On top of that, I don't feel like that the amount of work I listed above (probably a month worth?) is worth putting into this tool, while it is useful, I don't feel anyone would really miss it if it disappeared overnight.

So now I'm officially stepping down as a maintainer/owner/whatever, I practically haven't been one for years anyway.

Thank you @panekj @ia0 @JounQin and everyone else who helped keep this alive.

My question is **what should we do with all of this?**
- transfer everything to an organization?
- keep the current state up with sporadic maintenance until the TOML spec changes and everything breaks?
- ultimately is there anyone who would like to pour time into this and take over?

The following needs to be thought of:
 - [ ] this repo
 - [ ] access to the extension on [openvsx](https://open-vsx.org/extension/tamasfe/even-better-toml)
 - [ ] access to the extension on the [vscode marketplace](https://marketplace.visualstudio.com/items?itemName=tamasfe.even-better-toml)
 - [ ] access to the `@taplo` organization on NPM
 - [ ] access the [`taplo`](https://pypi.org/project/taplo/) package on pypi (I don't have anything to do with this)

I'm open to anything really and suggestions are welcome.

**Comment by JounQin** [comment]
The orgnization permission model is much more flexible IMO.

**Comment by ia0** [comment]
Thanks @tamasfe for all the efforts and care you've put into this project. This makes it all harder to decide to step down.

While I'm still not planning to help maintain this project, I believe this is an important project for the Rust community (for `Cargo.toml` formatting). That's why I'll give my opinion to try to maximize the project chances of success.

I agree with both @panekj and @JounQin that a GitHub organization should be preferred. My rationale is that the set of maintainers is going to change over time and no maintainer is expected to survive the project. So a GitHub organization will provide the maximum stability to the project.

If we don't find an initial set of maintainers or if we don't reach a good enough quota of commitment, we could ask https://users.rust-lang.org/ for help.

**Comment by JP-Ellis** [comment]
Thank you for everything @tamasfe! Taplo has been and still is a great project, even if there are things you wish you could improve.

While I have not personally contributed to this (yet), I would like to echo that moving to an organization is probably for the best.

To help attract new contributors, it would also be amazing to try and flesh out some of the ideas you have above. In my experience, creating an overarching tracking issue (like [I have done in Pact-Python](https://github.com/pact-foundation/pact-python/issues/396)) can provide a lot of help for new contributors to see what they can/should contribute to. This would really help pass on all the stuff you have thought of and help make it concrete for others to take over (though it is still a fair bit of work to create such a tracking issue).

**Comment by tamasfe** [comment]
Thank you all for the input, it looks like an organization is the way to go then. I will try to set one up as well as collect ideas/thoughts/TODOs like @JP-Ellis suggested for a proper handoff before I disappear (again).

**Comment by joshka** [comment]
I'd like to suggest also reaching out to the rust-analyzer owners to see if there's some scope to pull this into their responsibility area. While this isn't purely a Rust specific tool, there's quite a bit of overlap in the techniques and goals between rust-analyzer and this that could benefit from some cross pollination.

**Comment by panekj** [comment]
@tamasfe Also forgot to mention, but if there is anything I can help with to accelerate migration to GH org, please let me know.

**Comment by tamasfe** [comment]
Thank you, I think I'll just need a deep breath and a weekend, I wanted this done already, but might still take a while.

**Comment by yassun7010** [comment]
Hello. I am the author of [tombi](https://github.com/tombi-toml/tombi).

First of all, thank you @tomasfe for your great work.

I am currently developing a TOML Language Server inspired by taplo.

tombi is inspired by Rust Analyzer, black, and taplo,
It is already at the stage where it can be used in my team's product development.

Please use tombi and give me your feedback, and stars!

I am not sure if I will join with taplo, but I share tombi's status on this Issue.

**Comment by Mayeen-39** [comment]
Hi Tamás, hi everyone,

First, thank you @tamasfe for creating such a valuable tool. Even Better TOML and Taplo have become essential for many, including myself. I truly respect your transparency in this transition and the incredible work you've done bringing these tools to life with such polish.
I have been reaching out to Tamás directly over the past few weeks to discuss the possibility of taking over stewardship and ownership of the Even Better TOML VS Code extension. It is a tool I have been using and studying closely as part of an academic project, and it holds deep personal and professional importance to me. A few months ago, I started working on a project on a project with collaboration with an organization focused on open-source initiatives, and as someone committed to building tools for educational equity, I am very serious about the future of this tool.
I would like to express my interest in formally taking ownership and full maintenance responsibility of the VS Code extension, and if necessary, even the associated Taplo tooling.
 I am ready to own and maintain the plugin (Visual Studio Marketplace + Open VSX), keep it aligned with TOML updates, community needs, and performance expectations, modernize and document it with a focus on new contributors and student users.
And of course, offer compensation to facilitate this transition, in recognition of the groundwork laid and the responsibility involved.
If the idea of transferring the extension ownership is on the table, I would be grateful for the opportunity.

Thanks everyone.

**Comment by boozook** [comment]
> This project seems to be a good successor

I disagree, it useless for me, it has no formatting options for example.

**Comment by tamasfe** [comment]
This is quite unfortunate, I had the impression that Tombi had much better groundwork (and the codebase seemed much more reasonable as well).

I have been using taplo since the beginning as it fits my needs but I cannot deny the fact that it has a lot of bugs and the code is unmaintainable as-is, since an all-around refactor/rewrite is more time consuming than anyone is willing to take on, it is destined to eventually die.

@ya7010 would you be open to some collaboration in order incorporate taplo's features and configuration into tombi so that we can sunset this project?

**Comment by ya7010** [comment]
@titaniumtraveler 
I wasn't familiar with GitHub boards 🙏 
I changed it to public.

**Comment by tamasfe** [comment]
> > Whatever plans I had with this project do not really exist anymore, unfortunately I barely even do development outside work. I'm glad that what started as a learning project ended up being useful for many for this long. Thank you for your efforts keeping this alive, I hope you will overcome your health issues.
> 
> > This project seems to be a good successor, it incorporates most (if not all) improvements I wanted to do here.
> > I think I'll just pull the trigger and deprecate every extension and archive the repository.
> 
> In that case I'm going to fork off and maintain own tree of taplo, as I would rather not use tombi (few reasons) and I'll leave the deprecating process to you. Thank you for everything.

Y'all aren't making this easier.

No need to fork it, I was under the impression that you were already tired of "keeping it alive". I'd be happy to transfer all ownership to you, but I'd really not be involved in any further decisions regarding the project because my presence (or rather lack of) does not matter here.

**Comment by snoweuph** [comment]
> [@panekj](https://github.com/panekj) if you're going to maintain this, is there anything blocking the move to the org you created?

any new regarding this @panekj?
I want to know if I can somehow support.

**Comment by panekj** [comment]
I can't do anything about this since I'm only permitted to push new commits or change CI tokens. My plan for now is to just keep merging and creating new releases here as long as I can.

**Comment by snoweuph** [comment]
I thought you wanted to move to the Org? So the only problem is that you cant do that easly?

Couldn't you just push the tree into the Org?
I could help porting the Old Issues over to the org by hand.

**Comment by panekj** [comment]
That's not what I care about because issues can be looked up on this repo, the problem is that I would have to point all distributions to new upstream repo for updating `taplo` package, crates.io and npm would required new crates/package publishing (under different name) and if I had to support VSC - setting up open-vsx/VS marketplace (or whatever it is called now).

I'm happy to help maintaining the project (even if very slowly) as long as it doesn't create much burden on me, like for example: I'm ok with trying to fix issues for VSCode extension if people are using it and it's mostly me pushing small code fixes, but I don't want to deal with maintaining extension if I have to set it up myself as I'm not even a VS Code user anymore (I use and maintain things for Zed/Lapce, having to care for 3rd editor would be too much).

**Comment by snoweuph** [comment]
fair enough, so basicly with all the remotes and co, the only option would be to rename it when moving it and seting up all those remotes from scratch OR, have @tamasfe help with this once?

**Comment by panekj** [comment]
Correct, this was my motivation for making last release (at least for VSCode, since the token expired somewhen around in December)

**Comment by snoweuph** [comment]
Im neither a VSCode user (neovim), so I cant help with vscode really, what would be needed if renaming or moving to sub paths would be people to test for all the extensions/integrations, that actually use them.
